src/merkle_proof/poseidon_branch.rs

```diff
@@ -0,0 +1,134 @@
+}
+
+/// Provides a conversion between Branch and PoseidonBranch.
+///
+/// We extract the data from the `Branch` and store it appropiately
+/// inside of the `PoseidonBranch` structure with the bitflags already
+/// computed and the offsets pointing to the next levels pointing also to
+/// the correct places.
+impl<C, H> From<&Branch<'_, C, H>> for PoseidonBranch
+where
+    C: Compound<H>,
+    C::Annotation: Borrow<StorageScalar>,
+    H: ByteHash,
+{
+    fn from(branch: &Branch<C, H>) -> PoseidonBranch {
+        let mut poseidon_branch = PoseidonBranch::with_capacity(branch.levels().len() - 1);
+        // Skip root and store it directly.
+        poseidon_branch.root = branch
+            .levels()
+            .first()
+            .unwrap()
+            .annotation()
+            .unwrap()
+            .borrow()
+            .to_owned()
+            .into();
+        // Store the levels with the bitflags already computed inside
+        // of our PoseidonBranch structure.
+        for level in branch.levels().iter().rev() {
+            // Generate a default mutable `PoseidonLevel`, add the corresponding data
+            // extracted from the `Branch` and push it to our poseidon branch previously
+            // generated.
+            poseidon_branch.levels.push({
+                let mut pos_level = PoseidonLevel::default();
+                let mut level_bitflags = 0u64;
+                level
+                    .children()
+                    .iter()
+                    // Copy in poseidon_branch the leave values of the actual level with an
+                    // offset of one. So then we can add the bitflags at the beggining as the
+                    // first item of the `WIDTH` ones.
+                    .zip(pos_level.leaves.iter_mut().skip(1))
+                    // If they're null, place a Scalar::zero() inside of them as stated on the
+                    // Poseidon Hash paper.
+                    .enumerate()
+                    .for_each(|(idx, (src, dest))| {
+                        *dest = match src.annotation() {
+                            Some(borrow) => {
+                                let annotation: &Scalar = &(*borrow).borrow().borrow();
+                                // If the Annotation contains a value, we set the bitflag to 1.
+                                // Since the first element will be the most significant bit of the
+                                // bitflags, we need to shift it according to the `ARITY`.
+                                //
+                                // So for example:
+                                // A level with: [Some(val), None, None, None] should correspond to
+                                // Bitflags(1000). This means that we need to shift the first element
+                                // by `ARITY` and lately decrease the shift order by `idx`.
+                                level_bitflags += 1u64 << ((ARITY - 1) - idx);
```

Comment:
If we're dealing with bitflag, shouldn't be `|=` for clarity (I'm pretty sure the compiler doesn't care much).
Also, the comment does not explain the `- 1`, it should be included.

Reply:
I'll address the `-1` nitpick.

Regarding the `|=` makes sense. If we deal with bits it's more accurate.

Reply:
Done.